src/app/services/api.service.ts

```diff
@@ -3,88 +3,129 @@ import { HttpClient, HttpHeaders, HttpErrorResponse } from '@angular/common/http
+  ) { }
 
-constructor(private http: HttpClient) { }
+  // Http Options
+  httpOptions = {
+    headers: new HttpHeaders({
+      'Content-Type': 'application/json'
+    })
+  }
 
-// Http Options
-httpOptions = {
-  headers: new HttpHeaders({
-    'Content-Type': 'application/json'
-  })
-}
 
-// Handle API errors
-handleError(error: HttpErrorResponse) {
-  if (error.error instanceof ErrorEvent) {
-    // A client-side or network error occurred. Handle it accordingly.
-    console.error('An error occurred:', error.error.message);
-  } else {
-    // The backend returned an unsuccessful response code.
-    // The response body may contain clues as to what went wrong,
-    console.error(
-      `Backend returned code ${error.status}, ` +
-      `body was: ${error.error}`);
+  // Handle API errors
+  handleError(error: HttpErrorResponse) {
+    if (error.error instanceof ErrorEvent) {
+      // A client-side or network error occurred. Handle it accordingly.
+      console.error('An error occurred:', error.error.message);
+    } else {
+      // The backend returned an unsuccessful response code.
+      // The response body may contain clues as to what went wrong,
+      // console.error(
+      //   `Backend returned code ${error.status}, ` +
+      //   `body was: ${error.error}`);
+      console.log(error);
+      if (error.status == 401) {
+        return throwError(
+          { 'code': 401 }
+        );
+      }
+    }
+    // return an observable with a user-facing error message
+    return throwError(
+      'Something bad happened; please try again later.');
+  };
+
+
+  // Get login user by token
+  me(token): Observable<User> {
+    let httpOptionsWithToken = {
+      headers: new HttpHeaders({
+        'Content-Type': 'application/json',
+        // TODO: test when token is invalid
+        'authorization': 'Bearer d' + token
+      })
+    }
+
+    return this.http
+      .get<User>(this.ENV.me(), httpOptionsWithToken)
+      .pipe(
+        retry(2),
+        catchError(this.handleError)
+      )
   }
-  // return an observable with a user-facing error message
-  return throwError(
-    'Something bad happened; please try again later.');
-};
-
-
-// Create a new item
-createItem(item): Observable<Student> {
-  return this.http
-    .post<Student>(this.base_path, JSON.stringify(item), this.httpOptions)
-    .pipe(
-      retry(2),
-      catchError(this.handleError)
-    )
-}
 
-// Get single student data by ID
-getItem(id): Observable<Student> {
-  return this.http
-    .get<Student>(this.base_path + '/' + id)
-    .pipe(
-      retry(2),
-      catchError(this.handleError)
-    )
-}
 
-// Get students data
-getList(): Observable<Student> {
-  return this.http
-    .get<Student>(this.base_path)
-    .pipe(
-      retry(2),
-      catchError(this.handleError)
-    )
-}
+  // Login
+  login(credential): Observable<User> {
+    return this.http
+      .post<User>(this.ENV.login(), JSON.stringify(credential), this.httpOptions)
```

Comment:
is this better to call constructor and get all env as struct

or just need to call a method and get result?